Protocol methods doc (#41744)
dotnet/docs: docs/azure/sdk/protocol-convenience-methods.md

---
title: Understand Azure SDK client library method types
description: Learn about the key differences between Azure SDK client library protocol and convenience methods
ms.topic: conceptual
ms.custom: devx-track-dotnet, engagement-fy23, devx-track-arm-template
ms.date: 06/24/2024
---

# Azure SDK for .NET protocol and convenience methods overview

> The Azure SDK client libraries provide an interface to Azure service REST APIs by translating method calls into into messages sent via the respective service protocol. For REST API services, this means sending HTTP requests and converting the responses into runtime types. In this article, you'll learn about the different types of methods exposed by the client libraries and explore their implementation patterns.

Reviewer: ```suggestion
The Azure SDK client libraries provide an interface to Azure service REST APIs by translating method calls into messages sent via the respective service protocol. For REST API services, this means sending HTTP requests and converting the responses into runtime types. In this article, you'll learn about the different types of methods exposed by the client libraries and explore their implementation patterns.
```

## Understand protocol and convenience methods

The Azure SDK for .NET client libraries can expose two different categories of methods to make requests to an Azure service:

- **Protocol methods** provide a thin wrapper around the underlying REST API for a corresponding Azure service. These methods map primitive input parameters to HTTP request values and return a raw HTTP response object.

- **Convenience methods** provide a convenience layer over the lower-level protocol layer to add support for the .NET type system and other benefits. Convenience methods accept primitives or .NET model types as parameters and map them to the body of an underlying REST API request. These methods also handle various details of request and response management to allow developers to focus on sending and receiving data objects, instead of lower-level concerns.

### Azure SDK client library dependency patterns

Protocol and convenience methods implement slightly different patterns based on the underlying package dependency chain of the respective library. The Azure SDK for .NET client libraries depend on one of two different core libraries:

> - [**Azure.Core**](/dotnet/api/overview/azure/core-readme) provides shared primitives, abstractions, and helpers for building modern Azure SDK client libraries. These libraries follow the [Azure SDK Design Guidelines for .NET](https://azure.github.io/azure-sdk/dotnet_introduction.html) and use package and namespace names prefixed with 'Azure', such as [`Azure.Storage.Blobs`](/dotnet/api/overview/azure/storage.blobs-readme).

Reviewer: ```suggestion
- [**Azure.Core**](/dotnet/api/overview/azure/core-readme) provides shared primitives, abstractions, and helpers for building modern Azure SDK client libraries. These libraries follow the [Azure SDK Design Guidelines for .NET](https://azure.github.io/azure-sdk/dotnet_introduction.html) and use package names and namespaces prefixed with *Azure*, such as [`Azure.Storage.Blobs`](/dotnet/api/overview/azure/storage.blobs-readme).
```

### [Protocol method](#tab/protocol-methods)

The following code uses a `ContentSafetyClient` to call the `AnalyzeText` protocol method:

:::code source="snippets/protocol-convenience-methods/AzureCoreProtocol/Program.cs":::

The preceding code demonstrates the following protocol method patterns:

- Uses the `RequestContent` type to supply data for the request body.
- Uses the `RequestContext` type to configure request options.
- Returns data using the `Response` type.
- Reads the `ContentStream` to access the response data.

> [!NOTE]
> The preceding code configures the `ClientErrorBehaviors.NoThrow` for the `RequestOptions`. This option prevents non-success service responses status codes from throwing an exception, which means the app code should manually handle the response status code checks.

### [Convenience method](#tab/convenience-methods)

The following code uses a `ContentSafetyClient` to call the `AnalyzeText` convenience method:

> :::code source="snippets/protocol-convenience-methods/AzureCoreConvenience/Program.cs":::

Reviewer: These code samples would be even easier to digest if you add the attribute to highlight the specific line where the convenience or protocol method is being called. That will immediately draw attention to the relevant line of code.

### [Convenience method](#tab/convenience-methods)

Consider the following code that uses a `ChatClient` to call the `CompleteChat` convenience method:

:::code source="snippets/protocol-convenience-methods/SCMConvenience/Program.cs":::

The preceding code demonstrates the following `System.ClientModel` convenience method patterns:

- Uses a standard C# primitive or model type as a parameter.
- Returns a `ClientResult` type that represents the result of the operation.

---

## Protocol and convenience method usage guidance

Although the Azure SDK for .NET client libraries provide the option to use either protocol or convenience methods, prioritize using convenience methods in most scenarios. Convenience methods are designed to improve the development experience and provide flexibility for authoring requests and handling responses. However, both method types can be used in your app as needed. Consider the following criteria when deciding which type of method to use:

> - Convenience methods:
>   - Enable you to work with more friendly method parameter and response types.
>   - Handle various low-level concerns and optimizations for you.
> 
> - Protocol methods:
>   - Provide access to lower-level types, such as `RequestContext` and `RequestOptions`, which aren't available through convenience methods.
>   - Enable access to features of the underlying REST APIs that convenience methods don't expose.
>   - Enable you to create your own convenience methods around service endpoints that don't already have convenience methods.

Reviewer: ```suggestion
Convenience methods:
- Enable you to work with more friendly method parameter and response types.
- Handle various low-level concerns and optimizations for you.

Protocol methods:
- Provide access to lower-level types, such as `RequestContext` and `RequestOptions`, which aren't available through convenience methods.
- Enable access to features of the underlying REST APIs that convenience methods don't expose.
- Enable you to create your own convenience methods around service endpoints that don't already have convenience methods.
```

## See also

- [Understanding the Azure Core library for .NET](https://devblogs.microsoft.com/azure-sdk/understanding-the-azure-core-library-for-net/)

> - [System.ClientModel library for .NET](/dotnet/api/overview/azure/system.clientmodel-readme)

Reviewer: ```suggestion
- [Azure.Core library for .NET](/dotnet/api/overview/azure/core-readme)
- [System.ClientModel library for .NET](/dotnet/api/overview/azure/system.clientmodel-readme)
```

- Reads `response.Content` to access the response data.

> [!NOTE]
> The preceding code configures the `ClientErrorBehaviors.NoThrow` for the `RequestOptions`. This option prevents non-success service responses status codes from throwing an exception, which means the app code should manually handle the response status code checks.

---

### Libraries that depend on System.ClientModel

Some client libraries that connect to non-Azure services use patterns similar to the libraries that depend on `Azure.Core`. For example, the [`OpenAI`](https://www.nuget.org/packages/OpenAI/2.0.0-beta.7) library provides a client that connects to the OpenAI services. These libraries are based on a library called `System.ClientModel` that has patterns similar to `Azure.Core`.

### [Convenience method](#tab/convenience-methods)

Consider the following code that uses a `ChatClient` to call the `CompleteChat` convenience method:

:::code source="snippets/protocol-convenience-methods/SCMConvenience/Program.cs" highlight="10,11":::

The preceding code demonstrates the following `System.ClientModel` convenience method patterns:

- Uses a standard C# primitive or model type as a parameter.

> - Returns a `ClientResult` type that represents the result of the operation.

Reviewer 2: `ClientResult<T>` has the same implicit cast to `T` that Azure.Core's `Response<T>` does, so you should be able to write code similar to the Azure.Core-based sample where it's more like:

```diff
- ClientResult<ChatCompletion> completion = chatClient.CompleteChat("What is Azure?");
+ ChatCompletion completion = chatClient.CompleteChat("What is Azure?");
```

### [Protocol method](#tab/protocol-methods)

The following code uses a `ChatClient` to call the `CompleteChat` protocol method:

:::code source="snippets/protocol-convenience-methods/SCMProtocol/Program.cs" highlight="26-31":::

The preceding code demonstrates the following `System.ClientModel` protocol method patterns:

- Uses the `BinaryContent` type as a parameter to supply data for the request body.
- Uses the `RequestContext` type to configure request options.
- Returns data using the `ClientResult` type.
- Calls the `GetRawResponse` method to access the response data.

> > [!NOTE]

Reviewer 2: Looks like something may be missing from the formatting here?

![image](https://github.com/user-attachments/assets/bf1c972d-e35d-4bfc-b77f-0786277c38f2)

---
title: Understand Azure SDK client library method types
description: Learn about the key differences between Azure SDK client library protocol and convenience methods
ms.topic: conceptual
ms.custom: devx-track-dotnet, engagement-fy23, devx-track-arm-template
ms.date: 06/24/2024
---

# Azure SDK for .NET protocol and convenience methods overview

> The Azure SDK client libraries provide an interface to Azure service REST APIs by translating method calls into messages sent via the respective service protocol. For REST API services, this means sending HTTP requests and converting the responses into runtime types. In this article, you'll learn about the different types of methods exposed by the client libraries and explore their implementation patterns.

Reviewer 2: Did you mean?

```suggestion
The Azure SDK client libraries provide an interface to Azure services by translating method calls into messages sent via the respective service protocol. For REST API services, this means sending HTTP requests and converting the responses into runtime types. In this article, you'll learn about the different types of methods exposed by the client libraries and explore their implementation patterns.
```

### Azure SDK client library dependency patterns

Protocol and convenience methods implement slightly different patterns based on the underlying package dependency chain of the respective library. The Azure SDK for .NET client libraries depend on one of two different core libraries:

- [**Azure.Core**](/dotnet/api/overview/azure/core-readme) provides shared primitives, abstractions, and helpers for building modern Azure SDK client libraries. These libraries follow the [Azure SDK Design Guidelines for .NET](https://azure.github.io/azure-sdk/dotnet_introduction.html) and use package names and namespaces prefixed with *Azure*, such as [`Azure.Storage.Blobs`](/dotnet/api/overview/azure/storage.blobs-readme).

- [**System.ClientModel**](/dotnet/api/overview/azure/system.clientmodel-readme) is a core library that provides shared primitives, abstractions, and helpers for .NET service client libraries. The `System.ClientModel` library is a general purpose toolset designed to help build libraries for a variety of platforms and services, whereas the `Azure.Core` library is specifically designed for building Azure client libraries.

> [!NOTE]

> > The `Azure.Core` library itself also depends on `System.ClientModel` for various service building blocks. In the context of this article, the key differentiator for method patterns is whether a client library depends on `Azure.Core` or `System.ClientModel` directly, rather than through a transitive dependency.

Reviewer 2: nit:

```suggestion
> The `Azure.Core` library itself also depends on `System.ClientModel` for various client building blocks. In the context of this article, the key differentiator for method patterns is whether a client library depends on `Azure.Core` or `System.ClientModel` directly, rather than through a transitive dependency.
```

> > The `Azure.Core` library itself also depends on `System.ClientModel` for various service building blocks. In the context of this article, the key differentiator for method patterns is whether a client library depends on `Azure.Core` or `System.ClientModel` directly, rather than through a transitive dependency.

Reviewer 2: > In the context of this article, the key differentiator for method patterns is whether a client library depends on `Azure.Core` or `System.ClientModel` directly, rather than through a transitive dependency.

I believe Azure.AI.OpenAI doesn't have a direct dependency on SCM, but it is an SCM-based client.  It might be worth differentiating differently.  The simplest way might be to look at whether the client constructor takes ClientOptions that inherit from [Azure.Core.ClientOptions](https://learn.microsoft.com/en-us/dotnet/api/azure.core.clientoptions?view=azure-dotnet) or [System.ClientModel.Primitives.ClientPipelineOptions](https://learn.microsoft.com/en-us/dotnet/api/system.clientmodel.primitives.clientpipelineoptions?view=azure-dotnet-preview).  Although this may be a bit complex.

Another approach is to look at whether convenience methods return [Azure.Response\<T\>](https://learn.microsoft.com/en-us/dotnet/api/azure.response-1?view=azure-dotnet) or [System.ClientModel.ClientResult\<T\>](https://learn.microsoft.com/en-us/dotnet/api/system.clientmodel.clientresult-1?view=azure-dotnet-preview). Alternatively, users could check whether protocol methods return [Azure.Response](https://learn.microsoft.com/en-us/dotnet/api/azure.response?view=azure-dotnet) or [System.ClientModel.ClientResult](https://learn.microsoft.com/en-us/dotnet/api/system.clientmodel.clientresult?view=azure-dotnet-preview)

Author: @annelo-msft I'm trying to think of how to provide a succinct definition that is more conceptual rather than based on types in code examples. Should we just say something along the lines of:

"In the context of this article, the key differentiator for method patterns is whether a client library depends on `Azure.Core` or `System.ClientModel` directly, rather than through a transitive dependency. However, there are some exceptions to this rule, such as Azure.AI.OpenAI, which follows SCM patterns but depend on Azure.Core to gain access to certain types."

> > The `Azure.Core` library itself also depends on `System.ClientModel` for various service building blocks. In the context of this article, the key differentiator for method patterns is whether a client library depends on `Azure.Core` or `System.ClientModel` directly, rather than through a transitive dependency.

Reviewer 2: Just wanted to check -- are we concerned that `Azure.AI.OpenAI` has a direct dependency on Azure.Core but follows SCM-based client patterns?

Author: Resolving this one to continue the chat in the other thread.

|---------|---------|---------|
|Request body     | [`RequestContent`](/dotnet/api/azure.core.requestcontent)         | [`BinaryContent`](/dotnet/api/system.clientmodel.binarycontent)         |
|Advanced options     | [`RequestContext`](/dotnet/api/azure.requestcontext)         | [`RequestOptions`](/dotnet/api/system.clientmodel.primitives.requestoptions)         |
|Raw HTTP Response     | [`Response`](/dotnet/api/azure.response)         | [`PipelineResponse`](/dotnet/api/system.clientmodel.primitives.pipelineresponse)         |
|Return type with output model     | [`Response<T>`](/dotnet/api/azure.response-1)         | [`ClientResult<T>`](/dotnet/api/system.clientmodel.clientresult-1)         |

The sections ahead provide implementation examples of these concepts.

## Protocol and convenience method examples

The coding patterns and types used by client library protocol and convenience methods vary slightly based on whether the library depends on `Azure.Core` or `System.ClientModel`. The differences primarily influence the .NET types used for handling request and response data.

### Libraries that depend on Azure.Core

> Many Azure SDK client libraries depend on the `Azure.Core` library. For example, the [`Azure.AI.ContentSafety`](/dotnet/api/overview/azure/ai.contentsafety-readme) library depends on the `Azure.Core` library and provides a `ContentSafetyClient` class that exposes both protocol and convenience methods.

Reviewer 2: I think all of the "Track 2" libraries do -- @scottaddie, do you have a preference for how this is stated?  IIRC we are trying to phase out "Track 1" libraries. I don't feel strongly one way or another.

### [Protocol method](#tab/protocol-methods)

The following code uses a `ContentSafetyClient` to call the `AnalyzeText` protocol method:

> :::code source="snippets/protocol-convenience-methods/AzureCoreProtocol/Program.cs" highlight="18-20":::

Reviewer 2: super nit:

In the rendered view of the docs, it looks like protocol methods are in the tab that shows by default.  Given the general scariness of these methods, would it make sense to put the convenience method tab on top?

Author: I originally put Protocol methods as the default to increase awareness of what they are for those visiting the article, but I agree this probably makes more sense to imply that convenience methods are a more approachable default. Let's set convenience as the default.

Author: Also that's not a super nit, it's a good idea, and nits are fine :D

The preceding code demonstrates the following protocol method patterns:

- Uses the `RequestContent` type to supply data for the request body.
- Uses the `RequestContext` type to configure request options.
- Returns data using the `Response` type.

> - Reads the `ContentStream` to access the response data.

Reviewer 2: It just occurred to me that we're working to move away from using `response.ContentStream` in favor of using `response.Content` instead.  Would it be possible to update the sample to reflect that?

Reviewer 2: That would also make the SCM-based sample look more similar to the Azure.Core sample, which might be a nice side-effect.

The following code uses a `ContentSafetyClient` to call the `AnalyzeText` protocol method:

> :::code source="snippets/protocol-convenience-methods/AzureCoreProtocol/Program.cs" highlight="18-20":::

Reviewer: ```suggestion
:::code source="snippets/protocol-convenience-methods/AzureCoreProtocol/Program.cs" highlight="18-23":::
```

> > The preceding code configures the `ClientErrorBehaviors.NoThrow` for the `RequestOptions`. This option prevents non-success service responses status codes from throwing an exception, which means the app code should manually handle the response status code checks.

Reviewer 2: Would it make sense to show this in the sample?  e.g.

```diff
// Call the protocol method
Response response = safetyClient.AnalyzeText(
    prompt, 
    new RequestContext() { ErrorOptions = ErrorOptions.NoThrow});

+ if (response.Status != 200)
+ {
+  throw new RequestFailedException(response);
+ }

using (StreamReader streamReader = new StreamReader(response.ContentStream))
{
    // Display the results
    Console.WriteLine(streamReader.ReadToEnd());
}
```

Author: Wouldn't this imply that other 200s such as 202 would be an exception? I think that might be a bit strange as a sample, though I agree the current state is also a little odd. I added a note about handling status code, but we could also be more up front with a comment such as "This option is only set for example purposes" or something.

Reviewer 2: Yes, good observation.  The way our generators work is that in service API specifications, successful response codes are listed for each endpoint.  A service might say that it will return `200` or `202` as a success response, but say, `201` is missing from the list of success responses.  If for some reason a response with a `201` status code was returned, the client's convenience method would throw a `RequestFailedException` (or `ClientResultException` in SCM-based clients) because this is not a stated success code.  It is intended that protocol method users would read the service REST API docs to know which status codes correspond to those success codes listed by the service -- not just HTTP "success" status codes, e.g. 2xx codes.  This is a good example of what we mean when we say RequestContext and RequestOptions are low-level/advanced options types -- the user assumes a lot of responsibility when using them.

For the purposes of the sample, some suggestions include -- adding a comment to the above something like

```diff
+ // Any response code other than 200 from Azure ContentSafety AnalyzeText REST API is not considered a success response
+ // See REST API details at https://azure-ai-content-safety-api-docs.developer.azure-api.net/api-details#api=content-safety-service-2023-10-01&operation=TextOperations_AnalyzeText
+ if (response.Status != 200) 
+ {
+ throw new RequestFailedException(response); 
+ }
```

-- or maybe illustrating a conceptually simpler use of RequestContext/RequestOptions.  I'd suggest [AddHeader](https://learn.microsoft.com/en-us/dotnet/api/system.clientmodel.primitives.requestoptions.addheader?view=azure-dotnet-preview#system-clientmodel-primitives-requestoptions-addheader(system-string-system-string)), but unfortunately it's only on RequestOptions today.  Alternatively, there is [AddPolicy](https://learn.microsoft.com/en-us/dotnet/api/azure.requestcontext.addpolicy?view=azure-dotnet#azure-requestcontext-addpolicy(azure-core-pipeline-httppipelinepolicy-azure-core-httppipelineposition)) or a [CancellationToken](https://learn.microsoft.com/en-us/dotnet/api/azure.requestcontext.cancellationtoken?view=azure-dotnet#azure-requestcontext-cancellationtoken) property on both.

> > The preceding code configures the `ClientErrorBehaviors.NoThrow` for the `RequestOptions`. This option prevents non-success service responses status codes from throwing an exception, which means the app code should manually handle the response status code checks.

Reviewer 2: nit:

```suggestion
> The preceding code configures the `ClientErrorBehaviors.NoThrow` for the `RequestOptions`. This option prevents protocol methods from throwing an exception when the response status code does not correspond to a successful service response, which means the app code should manually handle the response status code checks.
```

- Uses a standard C# primitive or model type as a parameter.
- Returns a friendly C# type that represents the result of the operation.

---

### Libraries that depend on System.ClientModel

Some client libraries that connect to non-Azure services use patterns similar to the libraries that depend on `Azure.Core`. For example, the [`OpenAI`](https://www.nuget.org/packages/OpenAI/2.0.0-beta.7) library provides a client that connects to the OpenAI services. These libraries are based on a library called `System.ClientModel` that has patterns similar to `Azure.Core`.

### [Protocol method](#tab/protocol-methods)

The following code uses a `ChatClient` to call the `CompleteChat` protocol method:

:::code source="snippets/protocol-convenience-methods/SCMProtocol/Program.cs" highlight="26-31":::

The preceding code demonstrates the following `System.ClientModel` protocol method patterns:

- Uses the `BinaryContent` type as a parameter to supply data for the request body.
- Uses the `RequestContext` type to configure request options.
- Returns data using the `ClientResult` type.
- Calls the `GetRawResponse` method to access the response data.

> [!NOTE]

> > The preceding code configures the `ClientErrorBehaviors.NoThrow` for the `RequestOptions`. This option prevents non-success service responses status codes from throwing an exception, which means the app code should manually handle the response status code checks.

Reviewer 2: I really like how the other sample turned out!  Would it make sense to check the response code in this sample as well to show the correct use of `NoThrow` in both places?

Reviewer: For consistency reasons, I think Anne's suggestion makes a lot of sense.

Author: I'm not sure if I follow here - the OpenAI docs don't seem to define a specific set of status codes that are considered acceptable.  Do we just want to use the same code snippet but with a more general message about handling status codes?

Reviewer 2: I would suggest this is a docs bug for OAI.  They would need to specify this in their service API description, I think, in order to successfully generate the client.  Let me see if I can find it in that ...

Reviewer 2: Yep, it's listed in their service API description here: https://github.com/joseharriaga/openai-in-typespec/blob/main/.openapi3/openapi3-openai.yaml#L376

This is reflected in the AOAI service as well: https://learn.microsoft.com/en-us/azure/ai-services/openai/reference#responses-2

We could add a comment stating that this is true if there isn't a sufficiently official docs link to add to the comment in the sample.

:::code source="snippets/protocol-convenience-methods/SCMProtocol/Program.cs" highlight="26-31":::

The preceding code demonstrates the following `System.ClientModel` protocol method patterns:

- Uses the `BinaryContent` type as a parameter to supply data for the request body.
- Uses the `RequestContext` type to configure request options.
- Returns data using the `ClientResult` type.
- Calls the `GetRawResponse` method to access the response data.

> [!NOTE]

> > The preceding code configures the `ClientErrorBehaviors.NoThrow` for the `RequestOptions`. This option prevents non-success service responses status codes from throwing an exception, which means the app code should manually handle the response status code checks.

Reviewer: ```suggestion
> The preceding code configures the [ClientErrorBehaviors.NoThrow](/dotnet/api/system.clientmodel.primitives.clienterrorbehaviors) behavior for the `RequestOptions`. This option prevents non-success service responses status codes from throwing an exception, which means the app code should manually handle the response status code checks.
```

---

## Protocol and convenience method usage guidance

Although the Azure SDK for .NET client libraries provide the option to use either protocol or convenience methods, prioritize using convenience methods in most scenarios. Convenience methods are designed to improve the development experience and provide flexibility for authoring requests and handling responses. However, both method types can be used in your app as needed. Consider the following criteria when deciding which type of method to use:

Convenience methods:

- Enable you to work with more friendly method parameter and response types.
- Handle various low-level concerns and optimizations for you.

Protocol methods:

- Provide access to lower-level types, such as `RequestContext` and `RequestOptions`, which aren't available through convenience methods.
- Enable access to features of the underlying REST APIs that convenience methods don't expose.

> - Enable you to create your own convenience methods around service endpoints that don't already have convenience methods.

Reviewer 2: Given the conversation around reading the REST API docs, it might be helpful to add to this list something along those lines -- maybe "Require understanding the service's REST API documentation to handle requests and responses correctly.

---

> title: Understand Azure SDK protocol methods

Reviewer: ```suggestion
title: Understand Azure SDK client library method types
```

> description: Learn about the key differences between Azure SDK protocol methods and convenience methods

Reviewer: ```suggestion
description: Learn about the key differences between Azure SDK client library protocol and convenience methods
```

ms.topic: conceptual
ms.custom: devx-track-dotnet, engagement-fy23, devx-track-arm-template
ms.date: 06/24/2024
---

# Azure SDK for .NET protocol and convenience methods overview

> The Azure SDK client libraries provide an interface to Azure service REST APIs by translating method calls into underlying HTTP requests. In this article, you'll learn about the different types of methods exposed by the client libraries and explore their implementation patterns.

Reviewer 2: This is super nit-picky of me to say, but we have some Azure SDK clients that talk to non-HTTP services, which I think are not REST based.  But those ones don't have protocol methods (at least, at present we don't have a story around that).  I'm happy to let you rework to address however you think is best given that, but one suggestion might be to say something like "interfaces to Azure services by translating method calls into messages sent via the service protocol.  For REST API services, this means sending HTTP requests and turning the responses into runtime types."  ... or something?

## Understand protocol and convenience methods

The Azure SDK for .NET client libraries expose two different categories of methods to make requests to an Azure service:

- **Protocol methods** provide a thin wrapper around the underlying REST API for a corresponding Azure service. These methods map primitive input parameters to HTTP request values and return a raw HTTP response object.

- **Convenience methods** provide a convenience layer over the lower-level protocol layer to add support for the .NET type system and other benefits. Convenience methods accept primitives or C# model types as parameters and map them to the body of an underlying REST API request. These methods also handle various details of request and response management to allow developers to focus on sending and receiving data objects, instead of lower-level concerns.

### Azure SDK client library dependency patterns

> Protocol and convenience methods implement slightly different patterns based on the underlying package dependency chain of the respective library. The Azure SDK for .NET client libraries depend on one of two different lower-level libraries:

Reviewer 2: Might be worth calling them "core" libraries instead of lower-level libraries?  My thinking is that it's true they implement the lower-level part of the stack relative to client libraries, but they also contain convenience layer types like DefaultAzureCredential and other end-user types.

- [**Azure.Core**](/dotnet/api/overview/azure/core-readme) provides shared primitives, abstractions, and helpers for building modern Azure SDK for .NET client libraries. These libraries follow the [Azure SDK Design Guidelines for .NET](https://azure.github.io/azure-sdk/dotnet_introduction.html) and use package and namespace names prefixed with 'Azure', such as [`Azure.Storage.Blobs`](/dotnet/api/overview/azure/storage.blobs-readme).

> - [**System.ClientModel**](/dotnet/api/overview/azure/system.clientmodel-readme) is a lower-level library that provides shared primitives, abstractions, and helpers for .NET service client libraries. The `System.ClientModel` library is a general purpose toolset designed to help build libraries for a variety of platforms and services, whereas the `Azure.Core` library is specifically designed for building Azure client libraries.

Reviewer 2: It's also a core library ... the only real difference from Azure.Core is that the types don't have Azure branding, and have been designed so that they can be base types of branded types.

> [!NOTE]

> > The `Azure.Core` library itself also depends on the `System.ClientModel` for various service building blocks. In the context of this article, the key differentiator for method patterns is whether a client library depends on `Azure.Core` or `System.ClientModel` directly, rather than through a nested dependency.

Reviewer 2: nit: I have heard these called transitive dependencies, but I don't have 100% understanding of this area - maybe check?  If nested is commonly used and technically correct, then I will learn :)

> The following table compares some of the request and response types used by protocol and convenience methods based on whether the library depend on `Azure.Core` or `System.ClientModel`:

Reviewer: ```suggestion
The following table compares some of the request and response types used by protocol and convenience methods based on whether the library depends on `Azure.Core` or `System.ClientModel`:
```

|Request or response concern  |Azure.Core  | System.ClientModel |
|---------|---------|---------|
|Request body     | `RequestContent`         | `BinaryContent`         |
|Advanced options     | `RequestContext`         | `RequestOptions`         |

> |Raw HTTP Response     | `Response`         | `PipelineResponse`         |

Reviewer: Link to the corresponding API ref pages. For example, `PipelineResponse` should use this link: `/dotnet/api/system.clientmodel.primitives.pipelineresponse`.

> |Input model     | `IPersistableModel<T>`         | `IPersistableModel<T>`         |

Reviewer 2: Looking at this ... I'm not sure that all Azure clients have moved to implementing `IPersistableModel<T>` yet, although it is on the roadmap ... maybe we remove these lines for now?  (Sorry to have led you astray on that ...)

Also ❤️ the table 🙂

|Output model     | `IPersistableModel<T>`         | `IPersistableModel<T>`         |
|Return type with output model     | `Response<T>`         | `ClientResult<T>`         |

The sections ahead provide implementation examples of these concepts.

## Protocol and convenience method examples

> The coding patterns and types used by client library protocol and convenience methods vary slightly based on whether the library depends on `Azure.Core` or `System.ClientModel`. The differences primarily influence the C# types used for handling request and response data.

Reviewer: ```suggestion
The coding patterns and types used by client library protocol and convenience methods vary slightly based on whether the library depends on `Azure.Core` or `System.ClientModel`. The differences primarily influence the .NET types used for handling request and response data.
```

### Libraries that depend on Azure.Core

Many Azure SDK client libraries depend on the `Azure.Core` library. For example, the [`Azure.AI.ContentSafety`](/dotnet/api/overview/azure/ai.contentsafety-readme) library depends on the `Azure.Core` library and provides a `ContentSafetyClient` class that exposes both protocol and convenience methods.

> #### Work with protocol methods

Reviewer: What do you think about eliminating the "Work with protocol methods" and "Work with convenience methods" headings and replacing with "Protocol" and "Convenience" tabs? It would reduce the amount of vertical scrolling needed and therefore make this doc seem less intimidating.

The following code uses a `ContentSafetyClient` to call the `AnalyzeText` protocol method:

```csharp
// Create the client

> var safetyClient = new ContentSafetyClient(

Reviewer: Include the `using` statements too

Reviewer 2: Actually, I am not sure [ContentSafetyClient](https://learn.microsoft.com/en-us/dotnet/api/azure.ai.contentsafety.contentsafetyclient?view=azure-dotnet) implements IDisposable?

>     new Uri("content-safety-service-uri"),

Reviewer: ```suggestion
    new Uri("<content-safety-service-uri>"),
```

    new DefaultAzureCredential());

// Create the message content

> RequestContent message = RequestContent.Create(new

Reviewer: Let's import all these code samples from a sample app that we know builds. The infrastructure in the dotnet/docs repo will then ensure the sample compiles any time a change is made to the code.

    {
        text = "What is Microsoft Azure?",
    });

// Call the protocol method
Response response = safetyClient.AnalyzeText(
    message, 
    new RequestContext()
    {

>          ErrorOptions = ErrorOptions.NoThrow;

Reviewer 2: Setting `ErrorOptions.NoThrow` will mean that if a non-success service response status code is received the method won't throw `RequestFailedException`.   If we want to illustrate this in the sample, it would be good to call out that when users pass this option, it's their responsibility to check the status code to make sure the response is holding what they're expecting prior to using it.

    });

// Display the response data
using (StreamReader streamReader = new StreamReader(response.ContentStream))
{
    Console.WriteLine(streamReader.ReadToEnd());
}
```

The preceding code demonstrates the following protocol method patterns:

> - Uses the `RequestContent` type as a  to supply data for the request body.

Reviewer: Something is missing in this sentence

- Uses the `RequestContext` type to configure request options.
- Returns data using the `Response` type.
- Reads the `ContentStream` to access the response data.

#### Work with convenience methods

The following code uses a `ContentSafetyClient` to call the `AnalyzeText` convenience method:

```csharp
// Create the client
var safetyClient = new ContentSafetyClient(
    new Uri("content-safety-service-uri"),
    new DefaultAzureCredential());

// Call the convenience method
AnalyzeTextResult result = safetyClient.AnalyzeText("What is Microsoft Azure?");

// Display the response data
foreach (var item in result.CategoriesAnalysis)
{
    Console.Write($"{item.Category}: ");
    Console.Write(item.Severity);
    Console.WriteLine();
}
```

The preceding code demonstrates the following `Azure.Core` convenience method patterns:

- Uses a standard C# primitive or model type as a parameter.
- Returns a friendly C# type that represents the result of the operation.

### Libraries that depend on System.ClientModel

> Some Azure client libraries depend directly on the `System.ClientModel` library. For example, the [`OpenAI`](https://www.nuget.org/packages/OpenAI/2.0.0-beta.7) library depends on the `System.ClientModel` library and provides a `ChatClient` class that exposes both protocol and convenience methods.

Reviewer 2: > Some Azure client libraries depend directly on the `System.ClientModel` library.

We actually expect all of them to move to direct dependencies on SCM in the not-too-distant future as model types implement `IPersistableModel<T>`, since that and the model reader-writer types have all been defined in SCM.

> Some Azure client libraries depend directly on the `System.ClientModel` library. For example, the [`OpenAI`](https://www.nuget.org/packages/OpenAI/2.0.0-beta.7) library depends on the `System.ClientModel` library and provides a `ChatClient` class that exposes both protocol and convenience methods.

Reviewer 2: OpenAI actually isn't an Azure client library -- Azure.AI.OpenAI is because it talks to the Azure OpenAI service. But the OpenAI client library talks to the third-party OpenAI service.

Reviewer 2: Given this, another suggestion for framing this -- given that SCM is intended primarily for unbranded clients -- maybe something like ... _client libraries that talk to non-Azure services are being created that use patterns similar to the ones developed for Azure.Core-based libraries.  For example, the OpenAI client that talks to the OpenAI services.  These libraries are based on a library called SCM that has patterns similar to Azure.Core._

I don't feel like I nailed that, but I am happy to let you run with it from there.

#### Work with protocol methods

The following code uses a `ChatClient` to call the `CompleteChat` protocol method:

```csharp
// Create the client
OpenAIClient client = new("your-openai-key");
ChatClient chatClient = client.GetChatClient("gpt-4");

// Create the request prompt content
BinaryData prompt = BinaryData.FromBytes("""
    {  
        "model": "gpt-4o",
        "messages": [
           {
               "role": "user",
               "content": "What is Microsoft Azure?."
           }
        ]
    }
    """u8.ToArray());
using BinaryContent content = BinaryContent.Create(prompt);

// Send the request
ClientResult result = chatClient.CompleteChat(
        content,
        new RequestOptions()
        { 

>             ErrorOptions = ClientErrorBehaviors.NoThrow

Reviewer 2: Same comment about NoThrow here ... if the service returns a 4xx code, response.Content or JsonDocument.Parse would throw.

    .GetProperty("message"u8)
    .GetProperty("content"u8)
    .GetString();

Console.WriteLine(message);
```

The preceding code demonstrates the following `System.ClientModel` protocol method patterns:

- Uses the `BinaryContent` type as a parameter to supply data for the request body.
- Uses the `RequestContext` type to configure request options.
- Returns data using the `ClientResult` type.
- Calls the `GetRawResponse` method to access the response data.

#### Work with convenience methods

Consider the following code that uses a `ChatClient` to call the `CompleteChat` convenience method:

```csharp
// Create the client
OpenAIClient client = new("your-openai-key");
ChatClient chatClient = client.GetChatClient("gpt-4");

// Send the prompt
ClientResult<ChatCompletion> completion
    = chatClient.CompleteChat("What is Azure?");

// Display the response data
Console.WriteLine($"{completion.Value.Role}: {completion.Value.Content}");
```

The preceding code demonstrates the following `System.ClientModel` convenience method patterns:

- Uses a standard C# primitive or model type as a parameter.
- Returns a `ClientResult` type that represents the result of the operation.

> 

Reviewer 2: I defer to you @alexwolfmsft, on how deep you want to go and how much detail you want to include per target audience, so feel free to disregard this suggestion if it's not at the level you're thinking.  One thought at this point is that it might be worth pointing out here or elsewhere that from both Azure and SCM-based convenience methods, you can call GetRawResponse on the return type to get access to the HTTP response details.  But that from protocol methods, the Azure clients return the HTTP response details directly, whereas in SCM-based protocol methods, you also call GetRawResponse to get the HTTP response details.  Maybe too low-level or not worth calling attention to -- your call.

## Protocol and convenience method usage guidance

> Although the Azure SDK for .NET client libraries provide the option to use either protocol or convenience methods, in most scenarios developers should prioritize using convenience methods. Convenience methods are designed to improve the development experience and provide flexibility for authoring requests and handling responses. Consider the following criteria when deciding which type of method to use:

Reviewer: We should point out that it's not an all-or-nothing type of decision. You can mix these method types in your app, as needed.

- Convenience methods:
  - Enable you to work with more friendly method parameter and response types.
  - Handle various low-level concerns and optimizations for you.

- Protocol methods:
  - Provide access to lower-level types such as `RequestContext` and `RequestOptions` that are not available through convenience methods.

>   - Enable access to features of the underlying REST APIs that convenience methods do not expose.

Reviewer 2: I would qualify that they may in some cases do this ... and it may also be worth adding a third point, that they may also provide service methods corresponding to service endpoints that don't have convenience methods for them yet -- but that those convenience methods may be added in later versions of the library.  Although maybe those are both part of the same point?

## See also

- [Understanding the Azure Core library for .NET](https://devblogs.microsoft.com/azure-sdk/understanding-the-azure-core-library-for-net/)

> - [System.ClientModel library for .NET](/dotnet/api/overview/azure/system.clientmodel-readme)

Reviewer 2: Someday we will add "Announcing the System.ClientModel library"! 😁
